### ISSUE TITLE
Support PEP 712's new attribute assignment conversion

### ISSUE THREAD

**Issue by thejcannon** [microsoft/pyright#5307]
**Describe the bug**
In the super cool experimental [PEP 712](https://peps.python.org/pep-0712/) support, we'll need to support attribute assignment as well: https://github.com/python/peps/pull/3152

**To Reproduce**
Just try and assign to a dataclass field with a converter and not use the field's type (but a supported conversion type)

**Expected behavior**
No error

**Code or Screenshots**

```python
from dataclasses import dataclass, field


@dataclass
class Foo:
    x: str = field(default=1, converter=str)


f = Foo(1)
f.x = 1
#  /I'm/a/path/tofoo.py:10:7 - error: Cannot assign member "x" for type "Foo"
#   "Literal[1]" is incompatible with "str" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
`pyright 1.1.314` :pie: 

**Additional context**
<3 y'all

**Comment by erictraut** [comment]
My understanding is that the converter runs only on the values passed at initialization time. Is that not the case? The current PEP says nothing about converting values after initialization.

It's not immediately clear from the `attrs` documentation either, but I just verified through experimentation that it indeed does support conversion when assigning to an attribute with an associated field. I guess that means we need to support it here. I recommend adding more details to the PEP to make this clear.

I think the most straightforward (and accurate) way to handle this is to synthesize descriptor objects for each of the defined attributes. The `__set__` method for these descriptor objects will accept any type accepted by the converter. The `__get__` method would return the transformed type (the output of the converter).

@debonte, we can talk about the details later today, assuming that you want to implement this.

**Comment by thejcannon** [comment]
> My understanding is that the converter runs only on the values passed at initialization time. Is that not the case? The current PEP says nothing about converting values after initialization.

You got that right. The new behavior is in [the linked PR](https://github.com/python/peps/pull/3152) (here's the PEP link if you want fewer clicks: https://pep-previews--3152.org.readthedocs.build/pep-0712/)

And your modeling this as a descriptor is likely a good way to model the world. However, wouldn't the `__get__` method be better modeled as returning the annotated attribute type since that is a superset of the converter return type?

**Comment by erictraut** [comment]
Yes, the `__get__` method would return the annotated attribute type. The `__set__` would accept the same input type(s) as the converter. This will typically create what's known as an "asymmetric descriptor" — one whose setter and getter have different types. This is important because asymmetric descriptors are treated specially by type checkers for purposes of type narrowing. Normally, the type of an expression `foo.x` would be narrowed on assignment based on the assigned value type, but if `x` is an asymmetric descriptor, it should not be narrowed on assignment. If we model these as descriptors, we inherit all of this existing logic for asymmetric descriptor detection and handling.

**Comment by thejcannon** [comment]
I briefly thought about describing the behavior in terms of descriptors in the PEP, but thought better of it.

Glad to know I was "on the right track" in terms of modeling the behavior. Maybe I'll have a future home as a typist... type-interested-person? What do you call yourselves? :stuck_out_tongue:

**Comment by debonte** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
